File: Julia/Sudoku.jl
Coord = Int
Block = Int
State = Int
Count = Int
struct NoSolutionException <: Exception end
struct Sudoku
    state::Array{State, 1}
    count::Array{Count, 1}
    queue::Array{Coord, 1}
end

M = 3
N = M * M
N_COORD = N * N * N
N_BLOCK = 4 * N * N

OPEN   = 0
FIXED  = 1
BANNED = 2

CLOSED = N + 1

DIGITS = "123456789"

coord(i, j, k) = ((i - 1) * N + j - 1) * N + k
block(v, p, q) = ((v - 1) * N + p - 1) * N + q

parents = zeros(Block, N_COORD, 4)
for i in 1:N, j in 1:N, k in 1:N
    c = coord(i, j, k)
    p = div(i - 1, M) * M + div(j - 1, M) + 1
    parents[c, 1] = block(1, i, j)
    parents[c, 2] = block(2, i, k)
    parents[c, 3] = block(3, j, k)
    parents[c, 4] = block(4, p, k)
end

children = zeros(Coord, N_BLOCK, N)
for b in 1:N_BLOCK
    children[b, :] = find(any(parents .== b, 2))
end

Sudoku() = begin
    state = fill(OPEN, N_COORD)
    count = fill(N, N_BLOCK)
    queue = Coord[]
    Sudoku(state, count, queue)
end

Sudoku(str::String) = begin
    s = Sudoku()
    for i in 1:N, j in 1:N
        ij = (i - 1) * N + j
        if ij <= length(str)
            k = search(DIGITS, str[ij])
            if k != 0
                push!(s.queue, coord(i, j, k))
            end
        end
    end
    s
end

Sudoku(s::Sudoku) = begin
    Sudoku(copy(s.state), copy(s.count), copy(s.queue))
end

show(s::Sudoku) = begin
    arr = fill('.', N, N)
    for i in 1:N, j in 1:N, k in 1:N
        if s.state[coord(i, j, k)] == FIXED
            arr[j, i] = DIGITS[k]
        end
    end
    join(arr)
end

solve(str::String) = begin
    try
        show(search!(Sudoku(str)))
    catch ME
        if ME isa NoSolutionException
            "NO SOLUTION"
        else
            rethrow(ME)
        end
    end
end

search!(s::Sudoku) = begin

    while !isempty(s.queue)
        c = shift!(s.queue)
        if s.state[c] == OPEN
            fix!(s, c)
        elseif s.state[c] == FIXED
            continue
        elseif s.state[c] == BANNED
            throw(NoSolutionException())
        end
    end

    m, b = findmin(s.count)
    if m == CLOSED
        return s
    end

    for c in children[b, :]
        if s.state[c] == OPEN
            try
                ss = Sudoku(s)
                push!(ss.queue, c)
                return search!(ss)
            catch ME
                if !(ME isa NoSolutionException)
                    rethrow(ME)
                end
            end
        end
    end

    throw(NoSolutionException())
end

fix!(s::Sudoku, c::Coord) = begin
    s.state[c] = FIXED
    for b in parents[c, :]
        close!(s, b)
    end
end

close!(s::Sudoku, b::Block) = begin
    s.count[b] = CLOSED
    for c in children[b, :]
        if s.state[c] == OPEN
            ban!(s, c)
        end
    end
end

ban!(s::Sudoku, c::Coord) = begin
    s.state[c] = BANNED
    for b in parents[c, :]
        if s.count[b] != CLOSED
            countdown!(s, b)
        end
    end
end

countdown!(s::Sudoku, b::Block) = begin
    s.count[b] -= 1
    if s.count[b] == 0
        throw(NoSolutionException())
    elseif s.count[b] == 1
        for c in children[b, :]
            if s.state[c] == OPEN
                push!(s.queue, c)
            end
        end
    end
end

for str in eachline()
    println(solve(str))
end

File: Python/Sudoku.py
def main():
    s = Sudoku()
    while True:
        try:
            print(s.solve(input()))
        except EOFError:
            break


UNIT = 3
SIZE = UNIT * UNIT
DIGITS = '123456789'

COORDS = SIZE * SIZE * SIZE
BLOCKS = 4 * SIZE * SIZE

OPEN = 0
FIXED = 1
BANNED = 2

DONE = SIZE + 1

INITIAL_STATE = [OPEN for _ in range(COORDS)]
INITIAL_COUNT = [SIZE for _ in range(BLOCKS)]


def coord(i: int, j: int, k: int) -> int:
    return (i * SIZE + j) * SIZE + k


b2cs: list[list[int]] = [
    *[
        [coord(i, j, k) for k in range(SIZE)]
        for i in range(SIZE)
        for j in range(SIZE)
    ],
    *[
        [coord(i, j, k) for j in range(SIZE)]
        for i in range(SIZE)
        for k in range(SIZE)
    ],
    *[
        [coord(i, j, k) for i in range(SIZE)]
        for j in range(SIZE)
        for k in range(SIZE)
    ],
    *[
        [coord(ii * UNIT + i, jj * UNIT + j, k)
         for i in range(UNIT) for j in range(UNIT)]
        for ii in range(UNIT)
        for jj in range(UNIT)
        for k in range(SIZE)
    ],
]


c2bs: list[list[int]] = [[] for _ in range(COORDS)]
for b, cs in enumerate(b2cs):
    for c in cs:
        c2bs[c].append(b)


class NoSolutionException(Exception):
    pass


class Sudoku:
    def __init__(self) -> None:
        self.state = [*INITIAL_STATE]
        self.count = [*INITIAL_COUNT]
        self.queue = []
        self.cache = None

    def __str__(self) -> str:
        mat = [['.' for _ in range(SIZE)] for _ in range(SIZE)]
        for i in range(SIZE):
            for j in range(SIZE):
                for k in range(SIZE):
                    if self.state[coord(i, j, k)] == FIXED:
                        mat[i][j] = DIGITS[k]
        return ''.join(''.join(row) for row in mat)

    def solve(self, problem: str) -> str:
        self.reset()
        self.feed(problem)
        try:
            return str(self.search())
        except NoSolutionException:
            return 'NO SOLUTION'

    def reset(self) -> None:
        self.state[:] = INITIAL_STATE
        self.count[:] = INITIAL_COUNT
        self.queue.clear()

    def load(self, s: 'Sudoku') -> None:
        self.state[:] = s.state
        self.count[:] = s.count
        self.queue.clear()

    def feed(self, problem: str) -> None:
        for ij, ch in enumerate(problem[:(SIZE * SIZE)]):
            if ch in DIGITS:
                i, j = divmod(ij, SIZE)
                k = DIGITS.index(ch)
                self.queue.append(coord(i, j, k))

    def search(self) -> 'Sudoku':
        while self.queue:
            c0 = self.queue.pop()
            if self.state[c0] == FIXED:
                continue
            if self.state[c0] == BANNED:
                raise NoSolutionException()
            self.state[c0] = FIXED
            for b1 in c2bs[c0]:
                self.count[b1] = DONE
                for c2 in b2cs[b1]:
                    if c2 == c0:
                        continue
                    if self.state[c2] == BANNED:
                        continue
                    if self.state[c2] == FIXED:
                        assert False
                    self.state[c2] = BANNED
                    for b3 in c2bs[c2]:
                        if b3 == b1:
                            continue
                        if self.count[b3] == DONE:
                            assert False
                        self.count[b3] -= 1
                        if self.count[b3] == 0:
                            raise NoSolutionException()
                        if self.count[b3] == 1:
                            for c4 in b2cs[b3]:
                                if self.state[c4] == OPEN:
                                    self.queue.append(c4)

        bmin = None
        mincnt = DONE
        for b, cnt in enumerate(self.count):
            if cnt < mincnt:
                bmin = b
                mincnt = cnt

        if mincnt == DONE:
            return self

        if self.cache is None:
            self.cache = Sudoku()
        for c in b2cs[bmin]:
            if self.state[c] == OPEN:
                self.cache.load(self)
                self.cache.queue.append(c)
                try:
                    return self.cache.search()
                except NoSolutionException:
                    pass

        raise NoSolutionException()


if __name__ == '__main__':
    main()
